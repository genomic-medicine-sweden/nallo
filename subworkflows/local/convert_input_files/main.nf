include { SAMTOOLS_IMPORT } from '../../../modules/nf-core/samtools/import/main'
include { SAMTOOLS_FASTQ  } from '../../../modules/nf-core/samtools/fastq/main'

// This subworkflow converts input files between FASTQ and BAM formats,
// while also outputting any input files that were not converted
// into the respective output channels.
workflow CONVERT_INPUT_FILES {
    take:
    ch_input      // channel: [ val(meta), reads ]
    convert_bam   //    bool: Should BAM files be converted to FASTQ
    convert_fastq //    bool: Should FASTQ files be converted to BAM

    main:
    ch_versions = Channel.empty()

    ch_input
        .branch { _meta, reads ->
            fastq: reads.extension == 'gz'
            bam: reads.extension == 'bam'
        }
        .set { reads_to_convert }

    ch_bam = reads_to_convert.bam
    ch_fastq = reads_to_convert.fastq

    if (convert_bam) {
        SAMTOOLS_FASTQ(
            reads_to_convert.bam,
            false,
        )
        ch_versions = ch_versions.mix(SAMTOOLS_FASTQ.out.versions)

        // Mix converted files back in
        ch_fastq = ch_fastq.mix(SAMTOOLS_FASTQ.out.other)
    }
    if (convert_fastq) {
        SAMTOOLS_IMPORT(
            reads_to_convert.fastq
        )
        ch_versions = ch_versions.mix(SAMTOOLS_IMPORT.out.versions)

        // Mix converted files back in
        ch_bam = ch_bam.mix(SAMTOOLS_IMPORT.out.bam)
    }

    emit:
    bam      = ch_bam // channel: [ val(meta), bam ]
    fastq    = ch_fastq // channel: [ val(meta), fastq ]
    versions = ch_versions // channel: [ versions.yml ]
}
