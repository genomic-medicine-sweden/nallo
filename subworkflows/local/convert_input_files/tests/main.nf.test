nextflow_workflow {

    name "Test Workflow CONVERT_INPUT_FILES"
    script "../main.nf"
    workflow "CONVERT_INPUT_FILES"

    test("convert BAM to FASTQ") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test_bam' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true)
                    ],
                    [
                        [ id: 'test_fastq' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.fastq.gz', checkIfExists: true)
                    ]
                )
                input[1] = true
                input[2] = false
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 1 },
                { assert workflow.out.fastq.size() == 2 },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("convert FASTQ to BAM") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test_bam' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true)
                    ],
                    [
                        [ id: 'test_fastq' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.fastq.gz', checkIfExists: true)
                    ]
                )
                input[1] = false
                input[2] = true
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.fastq.size() == 1 },
                { assert workflow.out.bam.size() == 2 },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("convert BAM to FASTQ and FASTQ to BAM") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test_bam' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true)
                    ],
                    [
                        [ id: 'test_fastq' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.fastq.gz', checkIfExists: true)
                    ]
                )
                input[1] = true
                input[2] = true
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 2 },
                { assert workflow.out.fastq.size() == 2 },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("convert neither BAM to FASTQ nor FASTQ to BAM") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test_bam' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true)
                    ],
                    [
                        [ id: 'test_fastq' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.fastq.gz', checkIfExists: true)
                    ]
                )
                input[1] = false
                input[2] = false
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 1 },
                { assert workflow.out.fastq.size() == 1 },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("convert BAM to FASTQ and FASTQ to BAM -stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test_bam' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true)
                    ],
                    [
                        [ id: 'test_fastq' ],
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.fastq.gz', checkIfExists: true)
                    ]
                )
                input[1] = true
                input[2] = true
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 2 },
                { assert workflow.out.fastq.size() == 2 },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
