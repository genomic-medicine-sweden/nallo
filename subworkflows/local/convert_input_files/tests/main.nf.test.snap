{
    "convert FASTQ to BAM": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "test_fastq.bam:md5,028eb2162af5bc295b6888af0f732979"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "2": [
                    "versions.yml:md5,f6a244a66e7dc41043ae81b21857bda2"
                ],
                "bam": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "test_fastq.bam:md5,028eb2162af5bc295b6888af0f732979"
                    ]
                ],
                "fastq": [
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f6a244a66e7dc41043ae81b21857bda2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-31T10:36:12.536021508"
    },
    "convert BAM to FASTQ": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "test_bam_other.fastq.gz:md5,f9d9d405021d7a64eb59929e90aed241"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c688783029643ca3fd84ba740c99c2a7"
                ],
                "bam": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ]
                ],
                "fastq": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "test_bam_other.fastq.gz:md5,f9d9d405021d7a64eb59929e90aed241"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c688783029643ca3fd84ba740c99c2a7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-29T16:03:48.837045521"
    },
    "convert BAM to FASTQ and FASTQ to BAM": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "test_fastq.bam:md5,028eb2162af5bc295b6888af0f732979"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "test_bam_other.fastq.gz:md5,f9d9d405021d7a64eb59929e90aed241"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c688783029643ca3fd84ba740c99c2a7",
                    "versions.yml:md5,f6a244a66e7dc41043ae81b21857bda2"
                ],
                "bam": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "test_fastq.bam:md5,028eb2162af5bc295b6888af0f732979"
                    ]
                ],
                "fastq": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "test_bam_other.fastq.gz:md5,f9d9d405021d7a64eb59929e90aed241"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c688783029643ca3fd84ba740c99c2a7",
                    "versions.yml:md5,f6a244a66e7dc41043ae81b21857bda2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-29T16:06:40.375371334"
    },
    "convert neither BAM to FASTQ nor FASTQ to BAM": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "2": [
                    
                ],
                "bam": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ]
                ],
                "fastq": [
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-31T10:44:50.460485827"
    },
    "convert BAM to FASTQ and FASTQ to BAM -stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "test_fastq.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "test_bam_other.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c688783029643ca3fd84ba740c99c2a7",
                    "versions.yml:md5,f6a244a66e7dc41043ae81b21857bda2"
                ],
                "bam": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_PacBio_Revio.bam"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "test_fastq.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "fastq": [
                    [
                        {
                            "id": "test_bam"
                        },
                        "test_bam_other.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ],
                    [
                        {
                            "id": "test_fastq"
                        },
                        "/genomic-medicine-sweden/test-datasets/0af2aaefba648925012d0203f8ec52a5cf7ba1ce/testdata/HG002_ONT.fastq.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c688783029643ca3fd84ba740c99c2a7",
                    "versions.yml:md5,f6a244a66e7dc41043ae81b21857bda2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-29T16:16:49.102813896"
    }
}
